### ISSUE TITLE
Big Model Tutorial Techniques and Systems to Train and Serve Bigger Models

### ISSUE THREAD

**Issue by pentium3** [pentium3/sys_reading#259]
https://zhuanlan.zhihu.com/p/562741952

**Comment by pentium3** [comment]
https://icml.cc/virtual/2022/tutorial/18440

https://zhuanlan.zhihu.com/p/562741952

**Comment by pentium3** [comment]
https://alpa.ai/icml22-tutorial.html

https://sites.google.com/view/icml-2022-big-model

# Trends Driving Big Models - Ion Stoica

## intro

- as models become larger, hardware (compute && memory) capacity increasing slowly. 
- it's hard to fit the whole model to a single machine 

so we need to parallelize the model

## two ways to parallelize model

### 1. inter-operator parallelism: 

- partition model by layers/stages. different tensor operators run on different GPUs. 
- Pipeline execution on both forward and backward paths
- GPUs can be on the same machine or **different** machines  -> could lead to diff communication pattern

![image](https://github.com/pentium3/sys_reading/assets/7352163/5cd2cc6e-927c-4903-b05a-b41948e90988)

### 2. intra-operator parallelism: 

- partition one layer/stage. one operator can run across multiple GPUs

![image](https://github.com/pentium3/sys_reading/assets/7352163/98fc1988-063c-4c07-b95f-11e5ae59d991)


## Challenges

Huge optimization space:
- Data parallelism
- Model parallelism
    -  Inter-operator parallelism
    - Intra-operator parallelism

Growing diversity of specialized hardware (e.g., GPUs, TPUs)

Huge and growing diversity of NN architectures

**Comment by pentium3** [comment]
https://icml.cc/virtual/2022/tutorial/18440

# How to train and serve big models?

![image](https://github.com/pentium3/sys_reading/assets/7352163/a377d637-06d0-4e27-b91a-5a90206072d5)

Using parallelization.

Scope of the tutorial: Parallelization techniques and systems for big models

- Techniques to train and serve big models
    - data, operator, pipeline, expert, “3D” parallelism
    - inter-op and intra-op parallelism
- Available tools to start using big models now
    - big model libraries: Megatron-LM, DeepSpeed, XLA/GSPMD, Alpa
    - big model weights: OPT-175B, BLOOM
- New frontiers and opportunities in systems for big models
    - Auto-parallelization

**Comment by pentium3** [comment]
# Preliminaries

## Historical context and problem overview

### 1. Data parallelism with Parameter Server

allow asynchronously training

![image](https://github.com/pentium3/sys_reading/assets/7352163/e03bb4fa-e678-415a-924b-093001e82679)

however, many existing system like pytorch DDP use **all-reduce** to update parameter synchronously,
since communication between GPUs (within single machine) is fast enough

![image](https://github.com/pentium3/sys_reading/assets/7352163/c4b32eee-3787-4ab8-82b1-a369e620b718)


### 2. Computational Graph and Placement

TensorFlow: DL computation formulated as a dataflow graph

memory of one GPU is limited, so we need to **place** different node of dataflow graph to different devices(GPUs)

for big model, deciding placement is a hard problem.

## Background

### 1. distributed DL Computation

computation pattern of DL training:

```
for each epoch {
    for a batch of input data {
        ▽L(): forward propagation: compute prediction with model
        ▽L(): backward propagation: apply loss function and then compute gradient 
        f(): use gradient to update model weights.
    }
}
```

![image](https://github.com/pentium3/sys_reading/assets/7352163/0d00e719-5d62-41a9-861d-af190fce1f05)


### 2. Classic Views of ML Parallelisms: Data Parallelism / Model Parallelism

Data Parallelism: 
- partition input data. replicate model. assume whole model can fit a single device.
- general and precise.

Model Parallelism
- partition model, replicate data.
- but by partition model, it could be partitioning parameters OR computation function. **The view creates ambiguity for methods that neither partitions data nor the model computation.**

## New view (this tutorial): Inter-op parallelism / Intra-op parallelism

- built on two pillars: computational graph and device cluster
- This view is based on their computing characteristics.
- This view facilitates the development of new parallelism methods

### 1. formulate DL Computation as Graph

![image](https://github.com/pentium3/sys_reading/assets/7352163/6fda0d01-3cc4-4c28-bd59-974dcf320511)

components:
- L: a two layer MLE and with MSE loss
- ▽L(): compute gradient
- f(): update weights using simple sgd

3 steps:
- forward: compute MSE loss based on parameter and input data D
- backward: compute gradient of every parameter
- update weights using gradient 

### 2. formulate compute resources as Device Cluster

![image](https://github.com/pentium3/sys_reading/assets/7352163/4aae9224-cc21-4937-b0d7-11f0225cabdb)


### 3. Problem formulation: How to partition the computational graph on the device cluster?

![image](https://github.com/pentium3/sys_reading/assets/7352163/9a7a37ad-ec07-4412-9ce0-b96682b50d45)

![image](https://github.com/pentium3/sys_reading/assets/7352163/5975fc5a-d91e-47a8-877b-58820a914601)


### 4. compute pattern inside Intra- and Inter-op Parallelism

In the following example, which computes a matrix multiplication:

![image](https://github.com/pentium3/sys_reading/assets/7352163/7a8d4eab-663f-4eec-90e0-19e46e54bcd7)

partition strategy 1: Intra-op parallelism

![image](https://github.com/pentium3/sys_reading/assets/7352163/fd2412a0-fc5e-419a-92ed-fe87a2161824)


(1):
- X is replicated on dev1 and dev2
- dev1: result1_left = matmul(X, w1_left)
- dev2: result1_right = matmul(X, w1_right)

(2):
- dev1: result2_up = matmul(result1_left, w2_up)
- dev2: result2_down = matmul(result1_right, w2_down)
- result2 = np.sum(result2_up, result2_down)        [require collective all-reduce communication to aggregate results across all devices]

![image](https://github.com/pentium3/sys_reading/assets/7352163/1e641f8c-4119-41de-9de2-30aea3d162ec)


pattern of Collective Communication: (from NCCL documentation)

used **all-reduce** to aggregate partial results across all devices to a full tensor onto devices

![image](https://github.com/pentium3/sys_reading/assets/7352163/9a083850-34b9-4e40-aa95-2517f8be767c)


partition strategy 2: Inter-op parallelism

![image](https://github.com/pentium3/sys_reading/assets/7352163/6d41c0db-cc29-46a3-9ade-931ccb4ae598)

- need to send result1 from dev1 to dev2. only P2P communication
- dev2 is waiting while dev1 is running

pattern of Point-to-point Communication:

![image](https://github.com/pentium3/sys_reading/assets/7352163/5afaa323-85b2-48fe-b43c-9e86dc620513)



### 5. summarize:

![image](https://github.com/pentium3/sys_reading/assets/7352163/48669b94-3e5e-4bcf-aafb-a20628c61e5e)

we can formulate the ultimate problem with this new view of parallelism:

**What’s the best way to execute the graph subject to memory and communication constraints?**

**Comment by pentium3** [comment]
# Inter-op parallelism

suppose we partition a Computational Graph to 4 stages with inter-op parallelism, and put each stage on one device:

![image](https://github.com/pentium3/sys_reading/assets/7352163/f22c2067-efb0-4523-897b-098541b42d59)

## pipeline parallelism

### 1. pipeline bubble

due to data dependency, stage 2 can compute only after we get the result of stage 1.
if we visualize the execution timeline, we can see there are many times that devices are idle (gray area), a.k.a. Pipeline bubbles
- Pipeline bubble percentage = bubble_area / total_area

#### for Inference case
- to reduce pipeline bubble, use pipeline parallelism. eg, we feed 4 input (batches) a,b,c,d at one time in this example.
- Pipeline bubbles percentage = (D - 1) / (D - 1 + N) with D devices and N inputs 

![image](https://github.com/pentium3/sys_reading/assets/7352163/7f610be8-1658-4db5-94c4-42fea0b20634)


#### for Training
- harder to utilize pipeline parallelism, since we need to finish forward-backward-update path of one input before moving on to next input.
- Pipeline bubble percentage = (D - 1) / D, assuming D devices

![image](https://github.com/pentium3/sys_reading/assets/7352163/8f9c4887-3f50-4b4d-9914-7072097cd9f7)


### 2. How to Reduce Pipeline Bubbles for Training?

#### 2.1 Device Placement

Idea: Slice the branches of a neural network into multiple stages so they can be calculated concurrently

![image](https://github.com/pentium3/sys_reading/assets/7352163/453871cf-0f90-479c-863c-3d65a87739e7)

Limitation: (1) only work for NN with branch.    (2). dev utilization still low
device placement needs to be combined with the other pipeline schedules discussed later to further improve device utilization


#### 2.2 Synchronous Pipeline Parallel Schedule

Idea: Modify pipeline schedule to improve efficiency, but keep the computation and convergence semantics exactly the same as if training with a single device.

✅ Pros:
- Keep the convergence semantics. The training process is exactly the same as training the neural network on a single device.

❌ Cons:
- still have Pipeline bubbles.
- Reducing pipeline bubbles typically requires splitting inputs into smaller components, but too small input to the neural network will reduce the hardware efficiency.

**2.2.1 GPipe**

![image](https://github.com/pentium3/sys_reading/assets/7352163/82557227-2d93-4b1a-a24c-3df4b52bab87)

Limitation: memory usage for storing intermediate activation becomes higher as num of micro-batch increases

**2.2.2 1F1B:** 

optimize memory usage of GPipe by Perform backward as early as possible, so we don't need to keep all intermediate activation (of all mini-batches) in memory

![image](https://github.com/pentium3/sys_reading/assets/7352163/884796c9-7e2b-48e7-a0c2-91174f1e8c4c)


**2.2.3 Interleaved 1F1B**

![image](https://github.com/pentium3/sys_reading/assets/7352163/63167ece-8367-4dc2-874d-790cf1ac9224)

- Pro: Higher pipeline efficiency with fewer pipeline bubbles.
- Con: More communication overhead between stages


**2.2.4 TeraPipe**

apply pipeline parallelism for transformer model

![image](https://github.com/pentium3/sys_reading/assets/7352163/cd98eda8-5f5c-4f9d-8ad3-76689a4369c0)


**2.2.5 Chimera**

optimize 1F1B by inserting extra pipelines

![image](https://github.com/pentium3/sys_reading/assets/7352163/ca5d5701-7b28-41c2-8b1a-fc69a25ea15b)



#### 2.3 Asynchronous Pipeline Parallel Schedule

Idea: Start next round of forward pass **before** backward pass finishes.

✅ Pros:
- No Pipeline bubbles.

❌ Cons:
- Break the synchronous training semantics. Now the training will involve stalled gradient.
- Algorithms may store multiple versions of model weights for consistency.

**2.3.1 AMPNet**

different stages can have different version of weights. 

eg, in this example, suppose stage1 on dev1, stage2 on dev2, stage3 on dev3. 
For data#1, its forward path of stage 1 and 2 are using initial weights. But its forward path of stage 3 (and backward path of all stages) is using weights updated by data#0

it might bring noise to training process. hard to work for larger dataset

![image](https://github.com/pentium3/sys_reading/assets/7352163/0ed93b9d-2a13-4840-8bfe-6d6cd9afb2c2)

**2.3.2 Pipedream**

![image](https://github.com/pentium3/sys_reading/assets/7352163/fcdd3df2-d094-4f2a-928c-a4df714d9149)

#### 2.4 Automatic Stage Partitioning

The Assumption of previous Pipeline Parallel Schedule algorithms: assume balanced stages: running time (latency) of each stage are the same. Pipeline schedules works best with balanced stages.

![image](https://github.com/pentium3/sys_reading/assets/7352163/41f5618f-ccd5-42ab-abb8-57f02e159825)

however, Imbalance stage will create more pipeline bubble.
so, an important research problem is to **Minimize maximum stage latency & maximize parallelization**

there are generally two types of solutions:

**2.4.1 Reinforcement Learning Based (mainly for device placement):**

![image](https://github.com/pentium3/sys_reading/assets/7352163/d83ac896-e823-4eb3-817a-10862a7e4362)

**2.4.2 Optimization (Dynamic Programming/Linear Programming) Based:**

![image](https://github.com/pentium3/sys_reading/assets/7352163/f0da70ed-9ccb-4d6e-a154-f7fbcc8c752a)


#### 2.5 Summary

![image](https://github.com/pentium3/sys_reading/assets/7352163/e47ff751-1595-4dce-8dae-e8e4aa6a5250)

**Comment by pentium3** [comment]
# Intra-op parallelism

There are generally 2 ways to achieve Intra-op parallelism

## 1. parallelize a single operator

![image](https://github.com/pentium3/sys_reading/assets/7352163/f93873e4-2ef4-4663-9686-7bf401d977e4)

![image](https://github.com/pentium3/sys_reading/assets/7352163/287c585e-d3f5-4cad-b09d-1ab169db7114)

## 2. parallelize a subgraph

### a case of Intra-op Parallelism 

in the following example, we first look at a subgraph of the computation graph of a 2-layer MLP. 
there are 2 ways to partition input matrix to achieve intra-op parallelism:

![image](https://github.com/pentium3/sys_reading/assets/7352163/423f6fc0-01fa-43bb-82e8-78f890f306f1)

when we merge multiple subgraphs to the whole computation graph, since different operators’(in diff subgraph) parallelization strategies require different partition format of the same tensor, we need to consider re-partition Communication Cost:

![image](https://github.com/pentium3/sys_reading/assets/7352163/c4f1b72e-a1a4-4d10-9b0a-9472bfdd4332)

in this example, if subgraph 1 and 2 use different parallelization strategies, we need to re-partition the output of relu, which leads to some communication cost.

![image](https://github.com/pentium3/sys_reading/assets/7352163/a00c8651-1c1a-49ad-8c6c-ec9b47451be1)

### Problem formulation 

![image](https://github.com/pentium3/sys_reading/assets/7352163/6381d253-751b-4c24-95e2-92189231a207)

## 3. representative works using of Intra-op Parallelism 

### 3.1 Model-specific Intra-op Parallel Strategies

#### AlexNet

![image](https://github.com/pentium3/sys_reading/assets/7352163/c025c8f7-773e-4899-a65a-b542828e56e6)

#### Megaton-LM

weights are large, so we partition weights. lightweight operators like dropout are replicated.

![image](https://github.com/pentium3/sys_reading/assets/7352163/6129cf21-04cd-48c0-835b-5ae14b2b7050)

#### GShard MoE

partition the weight of MoE layers since it's large.

![image](https://github.com/pentium3/sys_reading/assets/7352163/7d69f0b8-1fdd-4bdd-aaee-19b918ffec06)

#### ZeRO Optimizer

![image](https://github.com/pentium3/sys_reading/assets/7352163/f6310065-15b0-41f5-a9ae-85299639ad45)

# 4. Combining inter- and intra-operator parallelism scales to more devices.

assign each stage to a device mesh (which contains multiple dev connected via diff mesh)

![image](https://github.com/pentium3/sys_reading/assets/7352163/3bf8b979-932b-42d6-9dd0-bc83bc29a8fb)

# 5. summary

- We can parallelize a single operator by exploiting its internal parallelism
- To do this for a whole computational graph, we need to choose strategies for all nodes in the graph to minimize the communication cost
- Intra-op and inter-op can be combined

**Comment by pentium3** [comment]
# New frontiers: auto-parallelization

X: best supported models of each system

Y: techniques used by each system

![image](https://github.com/pentium3/sys_reading/assets/7352163/5b9eb2ea-899d-4c3d-87ee-ef21f7a50f65)

**Auto-parallelization Problem: (automatically) find best combination of inter-op and intra-op strategies to maximize the performance of a model on a cluster of devices.**

![image](https://github.com/pentium3/sys_reading/assets/7352163/2d276996-fea5-4abf-b7b7-9909a59e6a81)

## challenges

The Search Space is Huge. 

- 100-10K num of operators in a real model (nodes to color)
- 80-200+ types if operators (type of nodes)
- 10-1000 num of devices in a cluster (available colors)

## existing Automatic Parallelization Methods

use 3 types of algorithms to find Parallelization strategy

![image](https://github.com/pentium3/sys_reading/assets/7352163/87942d79-81df-417a-b546-9f5187d889e1)

General Recipe of solving Automatic Parallelization problem:

![image](https://github.com/pentium3/sys_reading/assets/7352163/f8a8e4c8-54b1-4337-b54d-86c8fdbb25a6)

1. define a search space of diff strategies
2. reduce search space using heuristics
3. apply search algorithm to find best strategy
4. evaluate if the found strategy works well (by deploying strategy or human-designed cost model)
5. evaluation result can give feedback to search algorithm, if appliable  

## representative existing works

### 1. FlexFlow: use MCMC Search-based method

![image](https://github.com/pentium3/sys_reading/assets/7352163/8cadf4ec-55a6-412f-8665-61c58f793e87)

![image](https://github.com/pentium3/sys_reading/assets/7352163/8d3cf942-a279-4429-9fb3-f834efa5829c)

not practical with large model / cluster


### 2. ColocRL (Device Placement Optimization with Reinforcement Learning)

only consider inter-op


### 3. Alpa: Optimization-based Method

intuition: map inter-op(less communication but more device idle time) to slow connections in the cluster, map intra-op(more communication but less device idle time) to fast connections in the cluster. 
-> Minimize Computation cost + Communication cost

Hierarchically reduce search space: 
- first use inter-op to partition computation graph, and assign each stage to a subset of dev (use DP)
- then apply intra-op for each stage (use ILP)
- inter-op optimizer can call intra-op optimizer multiple times and get feedback

for more details see https://github.com/pentium3/sys_reading/issues/228


## comparison of 3 Automatic Parallelization Methods

![image](https://github.com/pentium3/sys_reading/assets/7352163/ab49ad0c-9a06-44cb-8ec1-b158aa342c9b)


![image](https://github.com/pentium3/sys_reading/assets/7352163/76a4c33d-3688-44b1-9a42-f5d6af259bd7)


## Future ML Systems and Challenges

automatic transform (scaling && scheduling) any prototype model code developed on a single machine, to a large scale parallel version that run on a cluster of devices, with achieving any performance goal.

![image](https://github.com/pentium3/sys_reading/assets/7352163/6b681c5b-5337-4eb9-a88f-59a3fab5b77d)

![image](https://github.com/pentium3/sys_reading/assets/7352163/74dc4c59-3fbf-4a1f-a133-ecbba8afbf53)
